libdrgn/program.c

```diff
@@ -788,20 +886,162 @@ struct drgn_error *drgn_program_find_prstatus_by_tid(struct drgn_program *prog,
+	struct drgn_thread *thread;
+	err = drgn_program_find_thread(prog, tid, &thread);
 	if (err)
 		return err;
 
-	it = drgn_prstatus_map_search(&prog->prstatus_map, &tid);
-	if (!it.entry) {
+	if (!thread) {
 		ret->str = NULL;
 		ret->len = 0;
 		return NULL;
 	}
-	*ret = it.entry->value;
+	*ret = thread->prstatus;
+	return NULL;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_find_thread(struct drgn_program *prog, uint32_t tid,
+			 struct drgn_thread **ret)
+{
+	struct drgn_error *err;
+
+	if (!(prog->flags & DRGN_PROGRAM_IS_LIVE)) {
+		err = drgn_program_cache_threads(prog);
+		if (err)
+			return err;
+	}
+
+	if (prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL) {
+		*ret = malloc(sizeof(**ret));
+		if (!*ret)
+			return &drgn_enomem;
+		(*ret)->prog = prog;
+		drgn_object_init(&(*ret)->thread, prog);
+
+		struct drgn_object init_pid_ns;
+		drgn_object_init(&init_pid_ns, prog);
+		err = drgn_program_find_object(prog, "init_pid_ns", NULL,
+					       DRGN_FIND_OBJECT_VARIABLE,
+					       &init_pid_ns);
+		if (err)
+			goto kernel_out;
+		err = drgn_object_address_of(&init_pid_ns, &init_pid_ns);
+		if (err)
+			goto kernel_out;
+		err = linux_helper_find_task(&(*ret)->thread, &init_pid_ns,
+					     tid);
+		if (err)
+			goto kernel_out;
+		(*ret)->tid = tid;
+	kernel_out:
+		drgn_object_deinit(&init_pid_ns);
+		if (err) {
+			drgn_object_deinit(&(*ret)->thread);
+			free(*ret);
+		}
+		return err;
+	} else {
+		*ret = drgn_thread_set_search(&prog->thread_set, &tid).entry;
+		return NULL;
+	}
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_crashed_thread(struct drgn_program *prog, struct drgn_thread **ret)
+{
+	struct drgn_error *err;
+
+	if (prog->flags & DRGN_PROGRAM_IS_LIVE)
+		return drgn_error_create(
+			DRGN_ERROR_INVALID_ARGUMENT,
+			"the crashed thread can only be found when debugging a core dump");
+	err = drgn_program_cache_threads(prog);
+	if (err)
+		return err;
+	if (!prog->crashed_thread)
+		return drgn_error_create(DRGN_ERROR_OTHER, "crashed thread not found");
+	*ret = prog->crashed_thread;
+	return NULL;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_thread_iterator_create(struct drgn_program *prog,
+			    struct drgn_thread_iterator **ret)
+{
+	struct drgn_error *err;
+	*ret = malloc(sizeof(**ret));
+	if (!*ret)
+		return &drgn_enomem;
+	struct drgn_object ns;
+	drgn_object_init(&ns, prog);
+	(*ret)->prog = prog;
+	if (prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL) {
+		drgn_object_init(&(*ret)->entry.thread, prog);
+		err = linux_helper_task_iter_init(&(*ret)->task_iterator, prog);
+	} else if (!(prog->flags & DRGN_PROGRAM_IS_LIVE)) {
+		err = drgn_program_cache_threads(prog);
+		if (err)
+			goto out;
+		(*ret)->iterator = drgn_thread_set_first(&prog->thread_set);
+	} else {
+		err = drgn_error_create(
+			DRGN_ERROR_INVALID_ARGUMENT,
+			"iterating over the threads of a live userpsace process is not currently supported");
+	}
+
+out:
+	drgn_object_deinit(&ns);
+	if (err)
```

Comment:
But right now in 2/3 cases it naturally reaches `out:` (there's only one `goto`), so structuring it similarly to the other example doesn't work as cleanly without introducing more otherwise unnecessary `if (err) goto out;`

Reply:
Fair enough.